Convert identity library to Kotlin.

davidz25 opened the pull request:
Also move all mdoc specific code to a new identity-mdoc library.

Test: Manually tested (appholder, appverifier, samples/preconsent-mdl, samples/age-verifier-mdl, wallet, secure-area-test-app)
Test: New unit tests and all unit tests pass (wallet, identity, identity-mdoc, identity-android).

dritan-x commented on appholder/src/main/java/com/android/identity/wallet/support/androidkeystore/AndroidAuthKeyCurveOption.kt:
```diff
@@ -1,18 +1,19 @@
-            X25519 -> SecureArea.EC_CURVE_X25519
+            P_256 -> EcCurve.P256
+            Ed25519 -> EcCurve.ED25519
+            X25519 -> EcCurve.X25519
```
I know your changes are about using the enum `EcCurve` instead of Ints, but it touches code that could also be refactored to be more Kotlin DSL 

```
    fun toEcCurve() = when (this) {
        P_256 -> EcCurve.P256
        Ed25519 -> EcCurve.ED25519
        X25519 -> EcCurve.X25519
    }
```

dritan-x commented on appholder/src/main/java/com/android/identity/wallet/support/softwarekeystore/SoftwareAuthKeyCurveOption.kt:
```diff
@@ -18,19 +19,19 @@ enum class SoftwareAuthKeyCurveOption : Parcelable {
     X448;
 
-    fun toEcCurve(): Int {
+    fun toEcCurve(): EcCurve {
```
same here

```
    fun toEcCurve() = when(this) {
        P256 -> EcCurve.P256
        ...
    }
```

dritan-x commented on identity-android/src/main/java/com/android/identity/android/mdoc/deviceretrieval/VerificationHelper.java:
```diff
@@ -1179,8 +1180,8 @@ KeyPair getEReaderKeyPair() {
-    public @SecureArea.EcCurve
-    int getEReaderKeyCurve() {
+    public
+    EcCurve getEReaderKeyCurve() {
```
syntax nit: `public` modifier can live in the same line as the rest of the method signature

davidz25 replied:
OK.

dritan-x commented on identity/src/main/java/com/android/identity/credential/AuthenticationKey.kt:
```diff
@@ -0,0 +1,229 @@
+     */
+    val attestation: List<X509Certificate>
+        get() {
+            return secureArea.getKeyInfo(alias).attestation
```
can be shortened to `get() = secureArea.getKeyInfo(alias).attestation`

dritan-x commented on identity/src/main/java/com/android/identity/credential/AuthenticationKey.kt:
```diff
@@ -0,0 +1,229 @@
+            credential: Credential
+        ): AuthenticationKey {
+            val ret = AuthenticationKey()
+            ret.alias = pendingAuthenticationKey.alias
```
using the `.apply{ .. }` [scope function](https://kotlinlang.org/docs/scope-functions.html#apply) we can make it neater 

```
        fun create(
            pendingAuthenticationKey: PendingAuthenticationKey,
            issuerProvidedAuthenticationData: ByteArray,
            validFrom: Timestamp,
            validUntil: Timestamp,
            credential: Credential
        ) = AuthenticationKey().apply {
            alias = pendingAuthenticationKey.alias
            domain = pendingAuthenticationKey.domain
            issuerProvidedData = issuerProvidedAuthenticationData
            this.validFrom = validFrom
            this.validUntil = validUntil
            this.credential = credential
            secureArea = pendingAuthenticationKey.secureArea
            privateApplicationData = pendingAuthenticationKey.privateApplicationData
            authenticationKeyCounter = pendingAuthenticationKey.authenticationKeyCounter
        }
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/CredentialRequest.kt:
```diff
@@ -0,0 +1,64 @@
+     * @param dataElementName the data element name.
+     * @param intentToRetain whether the remote reader intends to retain the data.
+     */
+    class DataElement(
```
should this be a [data class](https://kotlinlang.org/docs/data-classes.html) for easy readability of the instance parameters, comparison of instances, copy, etc..

davidz25 replied:
Probably, yea, doing that. We need to go back and refactor `CredentianRequest` once we have support for more than just mdoc.

dritan-x commented on identity/src/main/java/com/android/identity/credential/CredentialStore.kt:
```diff
@@ -0,0 +1,118 @@
+     * @param name name of the credential.
+     * @return A newly created credential.
+     */
+    fun createCredential(name: String): Credential {
+        val result = Credential.create(
+            storageEngine,
+            secureAreaRepository,
+            name
+        )
+        credentialCache[name] = result
+        return result
+    }
```
Kotlin DSL, using `also { .. }` [scope function](https://kotlinlang.org/docs/scope-functions.html#also) receives the resulting `Credential` instance in the also's context allowing you to perform some actions with the instance. The return value of `also { }` is the object for which the scope function was executed on.

```
    fun createCredential(name: String) = Credential.create(
        storageEngine,
        secureAreaRepository,
        name
    ).also {
        credentialCache[name] = it
    }
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/CredentialStore.kt:
```diff
@@ -0,0 +1,118 @@
+     * @param name the name of the credential.
+     * @return the credential or `null` if not found.
+     */
+    fun lookupCredential(name: String): Credential? {
+        var result = credentialCache[name]
+        if (result != null) {
+            return result
+        }
+        result = Credential.lookup(storageEngine, secureAreaRepository, name)
+        if (result == null) {
+            return null
+        }
+        credentialCache[name] = result
+        return result
+    }
```
Kotlin DSL

```
fun lookupCredential(name: String): Credential? {
        val result =
            credentialCache[name]
                ?: Credential.lookup(storageEngine, secureAreaRepository, name)
                ?: return null
        credentialCache[name] = result
        return result
    }
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/CredentialStore.kt:
```diff
@@ -0,0 +1,118 @@
+     * @return list of all credential names in the store.
+     */
+    fun listCredentials(): List<String> {
+        val ret = ArrayList<String>()
```
The Kotlin way is to declare a `mutableListOf<String>()` from [Kotlin Collections](https://kotlinlang.org/docs/collections-overview.html#list) -- which uses an `ArrayList` under the hood anyways 😂

davidz25 replied:
Ah, yea, I knew that, I missed that after going through the Java -> Kotlin mechanical conversion.

dritan-x commented on identity/src/main/java/com/android/identity/credential/CredentialUtil.kt:
```diff
@@ -0,0 +1,130 @@
+        for (authKey in credential.authenticationKeys) {
+            var keyExceededUseCount = false
+            var keyBeyondExpirationDate = false
+            if (authKey.domain != domain) {
+                // not one of ours...
+                continue
+            }
```
instead of doing this we can just filter for the correct domain to begin with. The for-loop above can look like so

```
 for (authKey in credential.authenticationKeys.filter { it.domain == domain }) {
```

davidz25 replied:
Even shorter: `val numExistingPendingKeys = credential.pendingAuthenticationKeys.filter { it.domain == domain }.size`

dritan-x commented on identity/src/main/java/com/android/identity/credential/CredentialUtil.kt:
```diff
@@ -0,0 +1,130 @@
+            numKeysNotNeedingReplacement++
+        }
+
+        var numExistingPendingKeys = 0
+        for (pendingAuthenticationKey in credential.pendingAuthenticationKeys) {
+            if (pendingAuthenticationKey.domain == domain) {
+                numExistingPendingKeys += 1
+            }
+        }
```
We can just count the results after applying `filter` in the new for-loop above (which infers that the variable be extracted outside of the for-loop so we can get the "numExistingPendingKeys"

```
val numExistingPendingKeys = existingPendingKeys.size
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+     */
+    val nameSpaceNames: List<String>
+        get() {
+            val ret: MutableList<String> = ArrayList()
```
can use `mutableListOf()` to simplify

`val ret = mutableListOf<String>()`

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+     * Names of all the namespaces.
+     */
+    val nameSpaceNames: List<String>
+        get() {
+            val ret: MutableList<String> = ArrayList()
+            for (nameSpaceName in map.keys) {
+                ret.add(nameSpaceName)
+            }
+            return ret
+        }
```
can simplified to a single line `get() = map.keys.toList()`

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+    fun getDataElementNames(nameSpaceName: String): List<String> {
+        val innerMap = map[nameSpaceName]
+            ?: throw IllegalArgumentException("No such namespace '$nameSpaceName'")
+        val ret: MutableList<String> = ArrayList()
+        for (dataElementName in innerMap.keys) {
+            ret.add(dataElementName)
+        }
+        return ret
```
these lines can be simplified to `return innerMap.keys.toList()`

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+     * @param dataElementName the data element name.
+     * @return `true` if there's a value for the data element, `false` otherwise.
+     */
+    fun hasDataElement(
+        nameSpaceName: String,
+        dataElementName: String
+    ): Boolean {
+        val innerMap = map[nameSpaceName] ?: return false
+        return innerMap[dataElementName] != null
+    }
```
More Kotlin DSL
```
fun hasDataElement(
        nameSpaceName: String,
        dataElementName: String
    ) = map[nameSpaceName]?.run { return this[dataElementName] != null } ?: false
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+    fun getDataElementString(
+        nameSpaceName: String,
+        dataElementName: String
+    ): String {
+        return Util.cborDecodeString(getDataElement(nameSpaceName, dataElementName))
+    }
```
can simply use "equal" without return

```
   fun getDataElementString(
        nameSpaceName: String,
        dataElementName: String
    ) = Util.cborDecodeString(getDataElement(nameSpaceName, dataElementName))
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+    fun getDataElementByteString(
+        nameSpaceName: String,
+        dataElementName: String
+    ): ByteArray {
+        return Util.cborDecodeByteString(getDataElement(nameSpaceName, dataElementName))
+    }
```
same as above 

```
 fun getDataElementByteString(
        nameSpaceName: String,
        dataElementName: String
    ) = Util.cborDecodeByteString(getDataElement(nameSpaceName, dataElementName))
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+    fun getDataElementNumber(
+        nameSpaceName: String,
+        dataElementName: String
+    ): Long {
+        return Util.cborDecodeLong(getDataElement(nameSpaceName, dataElementName))
+    }
```
```
fun getDataElementNumber(
        nameSpaceName: String,
        dataElementName: String
    ) = Util.cborDecodeLong(getDataElement(nameSpaceName, dataElementName))
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+    fun getDataElementBoolean(
+        nameSpaceName: String,
+        dataElementName: String
+    ): Boolean {
+        return Util.cborDecodeBoolean(getDataElement(nameSpaceName, dataElementName))
+    }
```
```
fun getDataElementBoolean(
        nameSpaceName: String,
        dataElementName: String
    ) = Util.cborDecodeBoolean(getDataElement(nameSpaceName, dataElementName))
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+                innerMapBuilder.put(UnicodeString(dataElementName), taggedBstr)
+            }
+        }
+        return builder.build().get(0)
```
can use `first()` instead of `get(0)`

`return builder.build().first()`

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+     * @return the bytes of the encoding describe above.
+     */
+    fun encodeAsCbor(): ByteArray {
+        return Util.cborEncode(toCbor())
```
"equals" without `return`

`fun encodeAsCbor() = Util.cborEncode(toCbor())`

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+ * This type is immutable.
+ */
+class NameSpacedData {
+    private var map: LinkedHashMap<String, LinkedHashMap<String, ByteArray>>
```
Kotlin DSL has `mutableMapOf`

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+class NameSpacedData {
+    private var map: LinkedHashMap<String, LinkedHashMap<String, ByteArray>>
+
+    internal constructor() {
+        map = LinkedHashMap()
+    }
+
+    private constructor(map: LinkedHashMap<String, LinkedHashMap<String, ByteArray>>) {
+        this.map = map
+    }
```
In Kotlin DSL, these 2 constructors don't need to exist since the `internal` constructor resorts to instantiating an empty map and the private one is called from the builder... these along with the class can be rewritten to 

```
class NameSpacedData private constructor(private val map: MutableMap<String, MutableMap<String, ByteArray>>) {
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+         * @param value the bytes of the CBOR.
+         * @return the builder.
+         */
+        fun putEntry(
+            nameSpaceName: String,
+            dataElementName: String,
+            value: ByteArray
+        ): Builder {
+            var innerMap = map[nameSpaceName]
+            if (innerMap == null) {
+                innerMap = LinkedHashMap()
+                map[nameSpaceName] = innerMap
+            }
+            innerMap[dataElementName] = value
+            return this
+        }
```
We can use `putIfAbsent` that only adds to the map if the specified key does not exist

```
fun putEntry(
            nameSpaceName: String,
            dataElementName: String,
            value: ByteArray
        ): Builder {
            map.putIfAbsent(nameSpaceName, mutableMapOf())
            map[nameSpaceName]?.run {
                this[dataElementName] = value
            }
            return this
        }
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+         * @param value A string.
+         * @return the builder.
+         */
+        fun putEntryString(
+            nameSpaceName: String,
+            dataElementName: String,
+            value: String
+        ): Builder {
+            return putEntry(nameSpaceName, dataElementName, Util.cborEncodeString(value))
+        }
```
```        
        fun putEntryString(
            nameSpaceName: String,
            dataElementName: String,
            value: String
        ) = putEntry(nameSpaceName, dataElementName, Util.cborEncodeString(value))
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+         * @param value A byte string.
+         * @return the builder.
+         */
+        fun putEntryByteString(
+            nameSpaceName: String,
+            dataElementName: String,
+            value: ByteArray
+        ): Builder {
+            return putEntry(nameSpaceName, dataElementName, Util.cborEncodeBytestring(value))
+        }
```
same as above

```
fun putEntryByteString(
            nameSpaceName: String,
            dataElementName: String,
            value: ByteArray
        ) = putEntry(nameSpaceName, dataElementName, Util.cborEncodeBytestring(value))
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+         * @param value The value, as a `long`.
+         * @return the builder.
+         */
+        fun putEntryNumber(
+            nameSpaceName: String,
+            dataElementName: String,
+            value: Long
+        ): Builder {
+            return putEntry(nameSpaceName, dataElementName, Util.cborEncodeNumber(value))
+        }
```
same as above 

```
        fun putEntryNumber(
            nameSpaceName: String,
            dataElementName: String,
            value: Long
        ) = putEntry(nameSpaceName, dataElementName, Util.cborEncodeNumber(value))
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+         * @param value The value, as a `boolean`..
+         * @return the builder.
+         */
+        fun putEntryBoolean(
+            nameSpaceName: String,
+            dataElementName: String,
+            value: Boolean
+        ): Builder {
+            return putEntry(nameSpaceName, dataElementName, Util.cborEncodeBoolean(value))
+        }
```
same as above 

```
        fun putEntryBoolean(
            nameSpaceName: String,
            dataElementName: String,
            value: Boolean
        ) = putEntry(nameSpaceName, dataElementName, Util.cborEncodeBoolean(value))
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+         *
+         * @return a [NameSpacedData] instance.
+         */
+        fun build(): NameSpacedData {
+            return NameSpacedData(map)
+        }
```
1-line function

`fun build() = NameSpacedData(map)`

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+         * [.encodeAsCbor].
+         */
+        @JvmStatic
+        fun fromEncodedCbor(encodedCbor: ByteArray): NameSpacedData {
+            return fromCbor(Util.cborDecode(encodedCbor))
+        }
```
1-liner

```
fun fromEncodedCbor(encodedCbor: ByteArray) = fromCbor(Util.cborDecode(encodedCbor))
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+        }
+
+        private fun fromCbor(dataItem: DataItem): NameSpacedData {
+            val ret = LinkedHashMap<String, LinkedHashMap<String, ByteArray>>()
```
this has to change to use `mutableMap`

`val ret = mutableMapOf<String, MutableMap<String, ByteArray>>()`

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+            for (nameSpaceNameItem in dataItem.keys) {
+                require(nameSpaceNameItem is UnicodeString) { "Expected string for namespace name" }
+                val namespaceName: String = nameSpaceNameItem.string
+                val dataElementToValueMap = LinkedHashMap<String, ByteArray>()
```
same here 
`val dataElementToValueMap = mutableMapOf<String, ByteArray>()`

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+                    val dataElementName: String = dataElementNameItem.string
+                    val valueItem: DataItem = dataElementItems[dataElementNameItem]
+                    require(valueItem is ByteString) { "Expected bytestring for data element value" }
+                    val value = valueItem.bytes
+                    dataElementToValueMap[dataElementName] = value
```
can simplify to 1 line 
`dataElementToValueMap[dataElementName] = valueItem.bytes`

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+                require(dataElementItems is Map) { "Expected map" }
+                for (dataElementNameItem in dataElementItems.keys) {
+                    require(dataElementNameItem is UnicodeString) { "Expected string for data element name" }
+                    val dataElementName: String = dataElementNameItem.string
```
no need to explicitly define type `String`

`val dataElementName: String = dataElementNameItem.string`

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+                for (dataElementNameItem in dataElementItems.keys) {
+                    require(dataElementNameItem is UnicodeString) { "Expected string for data element name" }
+                    val dataElementName: String = dataElementNameItem.string
+                    val valueItem: DataItem = dataElementItems[dataElementNameItem]
```
defining type `DataItem` is redundant
`val valueItem = dataElementItems[dataElementNameItem]`

dritan-x commented on identity/src/main/java/com/android/identity/credential/NameSpacedData.kt:
```diff
@@ -0,0 +1,386 @@
+                ret[namespaceName] = dataElementToValueMap
+            }
+            return NameSpacedData(ret)
+        }
```
all the changes above

```
 private fun fromCbor(dataItem: DataItem): NameSpacedData {
            val ret = mutableMapOf<String, MutableMap<String, ByteArray>>()
            require(dataItem is Map) { "dataItem is not a map" }
            for (nameSpaceNameItem in dataItem.keys) {
                require(nameSpaceNameItem is UnicodeString) { "Expected string for namespace name" }
                val namespaceName = nameSpaceNameItem.string
                val dataElementToValueMap = mutableMapOf<String, ByteArray>()
                val dataElementItems = Util.cborMapExtractMap(dataItem, namespaceName)
                require(dataElementItems is Map) { "Expected map" }
                for (dataElementNameItem in dataElementItems.keys) {
                    require(dataElementNameItem is UnicodeString) { "Expected string for data element name" }
                    val dataElementName = dataElementNameItem.string
                    val valueItem = dataElementItems[dataElementNameItem]
                    require(valueItem is ByteString) { "Expected bytestring for data element value" }
                    dataElementToValueMap[dataElementName] = valueItem.bytes
                }
                ret[namespaceName] = dataElementToValueMap
            }
            return NameSpacedData(ret)
        }
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/PendingAuthenticationKey.kt:
```diff
@@ -0,0 +1,227 @@
+         * @return An X.509 certificate chain for the pending authentication key.
+         */
+        get() {
+            return secureArea.getKeyInfo(alias).attestation
```
can be a 1-line function

```
        get() = secureArea.getKeyInfo(alias).attestation
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/PendingAuthenticationKey.kt:
```diff
@@ -0,0 +1,227 @@
+     * when this pending key was created.
+     */
+    val replacementFor: AuthenticationKey?
+        get() {
+            if (replacementForAlias == null) {
+                return null
+            }
+            for (authKey in credential.authenticationKeys) {
+                if (authKey.alias == replacementForAlias) {
+                    return authKey
+                }
+            }
+            Logger.w(TAG, "Key with alias $replacementForAlias which " +
+                    "is intended to be replaced does not exist")
+            return null
+        }
```
we can use `firstOrNull { }` to return `authKey` if there's a matching `authKey.alias`. 

To add logging we can use the `also` scope function

```
get() = credential.authenticationKeys.firstOrNull { it.alias == replacementForAlias }
            .also {
                if (it == null && replacementForAlias != null) {
                    Logger.w(
                        TAG, "Key with alias $replacementForAlias which " +
                                "is intended to be replaced does not exist"
                    )
                }
            }
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/PendingAuthenticationKey.kt:
```diff
@@ -0,0 +1,227 @@
+     * @param validUntil the point in time after which the data is not valid.
+     * @return a [AuthenticationKey].
+     */
+    fun certify(
+        issuerProvidedAuthenticationData: ByteArray,
+        validFrom: Timestamp,
+        validUntil: Timestamp
+    ): AuthenticationKey {
+        return credential.certifyPendingAuthenticationKey(
+            this,
+            issuerProvidedAuthenticationData,
+            validFrom,
+            validUntil
+        )
+    }
```
"equals" without `return`

```
fun certify(
        issuerProvidedAuthenticationData: ByteArray,
        validFrom: Timestamp,
        validUntil: Timestamp
    ) = credential.certifyPendingAuthenticationKey(
        this,
        issuerProvidedAuthenticationData,
        validFrom,
        validUntil
    )
```

dritan-x commented on identity/src/main/java/com/android/identity/credential/PendingAuthenticationKey.kt:
```diff
@@ -0,0 +1,227 @@
+        }
+        mapBuilder.put("applicationData", privateApplicationData.encodeAsCbor())
+            .put("authenticationKeyCounter", authenticationKeyCounter)
+        return builder.build().get(0)
```
can use `first()` instead of `get(0)`

dritan-x commented on identity/src/main/java/com/android/identity/credential/PendingAuthenticationKey.kt:
```diff
@@ -0,0 +1,227 @@
+    companion object {
+        const val TAG = "PendingAuthenticationKey"
+
+        fun create(
+            alias: String,
+            domain: String,
+            secureArea: SecureArea,
+            createKeySettings: CreateKeySettings,
+            asReplacementFor: AuthenticationKey?,
+            credential: Credential
+        ): PendingAuthenticationKey {
+            val ret = PendingAuthenticationKey()
+            ret.alias = alias
+            ret.domain = domain
+            ret.secureArea = secureArea
+            ret.secureArea.createKey(alias, createKeySettings)
+            if (asReplacementFor != null) {
+                ret.replacementForAlias = asReplacementFor.alias
+            }
+            ret.credential = credential
+            ret.privateApplicationData = SimpleApplicationData { ret.credential.saveCredential() }
+            ret.authenticationKeyCounter = credential.authenticationKeyCounter
+            return ret
+        }
```
can use the `run` scope function to simplify property accessors of `PendingAuthenticationKey`

```
        fun create(
            alias: String,
            domain: String,
            secureArea: SecureArea,
            createKeySettings: CreateKeySettings,
            asReplacementFor: AuthenticationKey?,
            credential: Credential
        ) = PendingAuthenticationKey().run {
            this.alias = alias
            this.domain = domain
            this.secureArea = secureArea
            this.secureArea.createKey(alias, createKeySettings)
            replacementForAlias = asReplacementFor?.alias
            this.credential = credential
            privateApplicationData = SimpleApplicationData { credential.saveCredential() }
            authenticationKeyCounter = credential.authenticationKeyCounter
        }
```

Note, for line 185, since `replacementForAlias` can be `null`, we can use `?` to do the checking and return either `null` or the value

davidz25 replied:
Took this but there's a missing `this` at the end of the `run` block, I think.

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+    }
+
+    @JvmStatic
+    fun toHex(bytes: ByteArray): String {
+        return toHex(bytes, 0, bytes.size)
+    }
```
1-liner function 

```
fun toHex(bytes: ByteArray): String = toHex(bytes, 0, bytes.size)
```

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+    }
+
+    @JvmStatic
+    fun base16(bytes: ByteArray): String {
+        return toHex(bytes).uppercase()
+    }
```
1-liner function

```
fun base16(bytes: ByteArray) = toHex(bytes).uppercase()
```

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+        return baos.toByteArray()
+    }
+
+    @JvmStatic
+    fun cborEncodeBoolean(value: Boolean): ByteArray {
+        return cborEncode(CborBuilder().add(value).build().get(0))
+    }
+
+    @JvmStatic
+    fun cborEncodeString(value: String): ByteArray {
+        return cborEncode(CborBuilder().add(value).build().get(0))
+    }
+
+    @JvmStatic
+    fun cborEncodeNumber(value: Long): ByteArray {
+        return cborEncode(CborBuilder().add(value).build().get(0))
+    }
+
+    @JvmStatic
+    fun cborEncodeBytestring(value: ByteArray): ByteArray {
+        return cborEncode(CborBuilder().add(value).build().get(0))
+    }
+
+    @JvmStatic
+    fun cborEncodeDateTime(timestamp: Timestamp): ByteArray {
+        return cborEncode(cborBuildDateTime(timestamp))
+    }
```
"equals" not `return`

```
    @JvmStatic
    fun cborEncodeBoolean(value: Boolean) =
        cborEncode(CborBuilder().add(value).build().get(0))

    @JvmStatic
    fun cborEncodeString(value: String) =
        cborEncode(CborBuilder().add(value).build().get(0))

    @JvmStatic
    fun cborEncodeNumber(value: Long) =
        cborEncode(CborBuilder().add(value).build().get(0))

    @JvmStatic
    fun cborEncodeBytestring(value: ByteArray) =
        cborEncode(CborBuilder().add(value).build().get(0))

    @JvmStatic
    fun cborEncodeDateTime(timestamp: Timestamp) =
        cborEncode(cborBuildDateTime(timestamp))
```

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+        require(dataItems.size == 1) {
+            "Unexpected number of items, expected 1 got ${dataItems.size}"
+        }
+        return dataItems.get(0)
```
can use `first()` instead of `get(0)` 
```
return dataItems.first()
```

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+
+    @JvmStatic
+    fun cborDecodeBoolean(data: ByteArray): Boolean {
+        val simple: SimpleValue
+        simple = try {
```
can join these 2 lines 

```
val simple: SimpleValue = try {
```

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+        return result
+    }
+
+    @JvmStatic
+    fun cborDecodeString(data: ByteArray): String {
+        return checkedStringValue(cborDecode(data))
+    }
```
"equals" not `return`

```
@JvmStatic
    fun cborDecodeString(data: ByteArray) =
        checkedStringValue(cborDecode(data))
```

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+     */
+    @JvmStatic
+    fun checkedStringValue(item: DataItem): String {
+        return castTo<UnicodeString, DataItem>(UnicodeString::class.java, item).string
```
the types of `castTo()` are inferred, also we can use "equals" and not `return`

```
    @JvmStatic
    fun checkedStringValue(item: DataItem) = castTo(UnicodeString::class.java, item).string
```

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+    }
+
+    @JvmStatic
+    fun cborDecodeLong(data: ByteArray): Long {
+        return checkedLongValue(cborDecode(data))
+    }
```
"equals" instead of `return`

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+    }
+
+    @JvmStatic
+    fun cborDecodeByteString(data: ByteArray): ByteArray {
+        val dataItem = cborDecode(data)
+        return castTo(
+            ByteString::class.java, dataItem
+        ).bytes
+    }
```
can put `cborDecode(data)` inside `castTo` and use "equals" instead of `return`

```
    fun cborDecodeByteString(data: ByteArray) =
        castTo(ByteString::class.java, cborDecode(data)).bytes
```

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+                // We're not the leaf certificate...
+                //
+                // Check the previous certificate was signed by this one.
+                try {
+                    prevCertificate.verify(certificate.publicKey)
+                } catch (e: CertificateException) {
+                    return false
+                } catch (e: InvalidKeyException) {
+                    return false
+                } catch (e: NoSuchAlgorithmException) {
+                    return false
+                } catch (e: NoSuchProviderException) {
+                    return false
+                } catch (e: SignatureException) {
+                    return false
+                }
```
can shorten this `try` with multiple `catches`

```
  try {
                    prevCertificate.verify(certificate.publicKey)
                } catch (e: Exception) {
                    when (e) {
                        is CertificateException,
                        is InvalidKeyException,
                        is NoSuchAlgorithmException,
                        is NoSuchProviderException,
                        is SignatureException -> return false
                    }
```

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+            verifier.initVerify(publicKey)
+            verifier.update(toBeSigned)
+            verifier.verify(derSignature)
+        } catch (e: SignatureException) {
+            throw IllegalStateException("Error verifying signature", e)
+        } catch (e: NoSuchAlgorithmException) {
+            throw IllegalStateException("Error verifying signature", e)
+        } catch (e: InvalidKeyException) {
+            throw IllegalStateException("Error verifying signature", e)
+        }
```
can combine these `catches`

```
            when (e) {
                is SignatureException,
                is NoSuchAlgorithmException,
                is InvalidKeyException ->
                    throw IllegalStateException("Error verifying signature", e)
                // some other exception was thrown
                else -> false
            }
```

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+        // Next field is the payload, independently of how it's transported (RFC
+        // 8152 section 4.4). Since our API specifies only one of |data| and
+        // |detachedContent| can be non-empty, it's simply just the non-empty one.
+        if (payload.size > 0) {
```
can use `isNotEmpty()`

```
payload.isNotEmpty()
```

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+        protectedHeadersMap.put(COSE_LABEL_ALG, COSE_ALG_HMAC_256_256)
+        val protectedHeadersBytes = cborEncode(protectedHeaders.build().get(0))
+        val toBeMACed = coseBuildToBeMACed(protectedHeadersBytes, data!!, detachedContent!!)
+        val mac: ByteArray
+        mac = try {
```
can join these 2 
```
 val mac: ByteArray = try {
```

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+            m.init(key)
+            m.update(toBeMACed)
+            m.doFinal()
+        } catch (e: NoSuchAlgorithmException) {
+            throw IllegalStateException("Unexpected error", e)
+        } catch (e: InvalidKeyException) {
+            throw IllegalStateException("Unexpected error", e)
+        }
```
can join catches

```
 } catch (e: Exception) {
            when (e) {
                is NoSuchAlgorithmException,
                is InvalidKeyException -> throw IllegalStateException("Unexpected error", e)
                else -> throw IllegalStateException("Unexpected error", e)
            }
        
```

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+     * Returns #6.24(bstr) of the given already encoded CBOR
+     */
+    @JvmStatic
+    fun cborBuildTaggedByteString(encodedCbor: ByteArray): DataItem {
+        val item: DataItem = ByteString(encodedCbor)
+        item.setTag(CBOR_SEMANTIC_TAG_ENCODED_CBOR)
+        return item
+    }
```
using `apply{ ..}` allows us to access the object and perform additions to it then return the object (with all the changes)

```
fun cborBuildTaggedByteString(encodedCbor: ByteArray): DataItem =
        ByteString(encodedCbor).apply { 
            setTag(CBOR_SEMANTIC_TAG_ENCODED_CBOR)
        }
```

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+    fun sec1EncodeFieldElementAsOctetString(
+        octetStringSize: Int,
+        fieldValue: BigInteger?
+    ): ByteArray {
+        return BigIntegers.asUnsignedByteArray(octetStringSize, fieldValue)
+    }
```
"equals" instead of `return`

```
    fun sec1EncodeFieldElementAsOctetString(
        octetStringSize: Int,
        fieldValue: BigInteger?
    ) = BigIntegers.asUnsignedByteArray(octetStringSize, fieldValue)
```

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+                    .put(COSE_KEY_PARAM_CRV, curve.coseCurveIdentifier.toLong())
+                    .put(COSE_KEY_PARAM_X, x)
+                    .end()
+                    .build().get(0)
```
`first()`

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+            val s = Signature.getInstance(algorithm)
+            s.initSign(key)
+            coseSign1Sign(s, data, additionalData, certificateChain)
+        } catch (e: NoSuchAlgorithmException) {
+            throw IllegalStateException("Caught exception", e)
+        } catch (e: InvalidKeyException) {
+            throw IllegalStateException("Caught exception", e)
+        }
```
I think the number of lines it takes to rewrite this in kotlin dsl is larger than having 2 catches thrown, this is a meritable tradeoff, either way works

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+                    .put(COSE_KEY_PARAM_X, x)
+                    .put(COSE_KEY_PARAM_Y, y)
+                    .end()
+                    .build().get(0)
```
.build().first()`

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+
+    @JvmStatic
+    fun cborMapHasKey(map: DataItem, key: String): Boolean {
+        val item = castTo<Map, DataItem>(
+            Map::class.java, map
+        )[UnicodeString(key)]
+        return item != null
```
the `castTo` types are inferred, we can then use "equals" instead of `return`

```
    fun cborMapHasKey(map: DataItem, key: String): Boolean =
        castTo(
            Map::class.java, map
        )[UnicodeString(key)] != null
```

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+    }
+
+    @JvmStatic
+    fun cborMapHasKey(map: DataItem, key: Long): Boolean {
+        val keyDataItem: DataItem = if (key >= 0) UnsignedInteger(key) else NegativeInteger(key)
+        val item = castTo(
+            Map::class.java, map
+        )[keyDataItem]
+        return item != null
```
can `return castTo(..) != null'
```
    fun cborMapHasKey(map: DataItem, key: Long): Boolean {
        val keyDataItem: DataItem = if (key >= 0) UnsignedInteger(key) else NegativeInteger(key)
        return castTo(
            Map::class.java, map
        )[keyDataItem] != null
    }
```

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+                ECPublicKeySpec(ecPoint, ecParameters)
+            val kf = KeyFactory.getInstance("EC")
+            kf.generatePublic(keySpec) as ECPublicKey
+        } catch (e: NoSuchAlgorithmException) {
+            throw IllegalStateException("Unexpected error", e)
+        } catch (e: InvalidParameterSpecException) {
+            throw IllegalStateException("Unexpected error", e)
+        } catch (e: InvalidKeySpecException) {
+            throw IllegalStateException("Unexpected error", e)
+        } catch (e: NoSuchProviderException) {
+            throw IllegalStateException("Unexpected error", e)
```
rewrite with a kotlin dsl approach

```
        } catch (e: Exception) {
            when (e) {
                is NoSuchAlgorithmException,
                is InvalidParameterSpecException,
                is InvalidKeySpecException,
                is NoSuchProviderException ->
                    throw IllegalStateException("Unexpected error", e)
                else ->
                    throw IllegalStateException("Unexpected error", e)
            }
        }
```

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+            baos.write(prefix)
+            baos.write(encodedX)
+            kf.generatePublic(X509EncodedKeySpec(baos.toByteArray()))
+        } catch (e: NoSuchAlgorithmException) {
+            throw IllegalStateException("Unexpected error", e)
+        } catch (e: InvalidKeySpecException) {
+            throw IllegalStateException("Unexpected error", e)
+        } catch (e: IOException) {
+            throw IllegalStateException("Unexpected error", e)
+        } catch (e: NoSuchProviderException) {
+            throw IllegalStateException("Unexpected error", e)
+        }
```
Exceptions can be joined

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+                ECPublicKeySpec(ecPoint, ecParameters)
+            val kf = KeyFactory.getInstance("EC")
+            kf.generatePublic(keySpec) as ECPublicKey
+        } catch (e: NoSuchAlgorithmException) {
+            throw IllegalStateException("Unexpected error", e)
+        } catch (e: InvalidParameterSpecException) {
+            throw IllegalStateException("Unexpected error", e)
+        } catch (e: InvalidKeySpecException) {
+            throw IllegalStateException("Unexpected error", e)
+        }
```
"join catches" like above

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+            val privateKeySpec = ECPrivateKeySpec(s, ecParameters)
+            val keyFactory = KeyFactory.getInstance("EC")
+            keyFactory.generatePrivate(privateKeySpec)
+        } catch (e: NoSuchAlgorithmException) {
+            throw IllegalStateException(e)
+        } catch (e: InvalidParameterSpecException) {
+            throw IllegalStateException(e)
+        } catch (e: InvalidKeySpecException) {
+            throw IllegalStateException(e)
```
"catches can be joined"

dritan-x commented on identity/src/main/java/com/android/identity/internal/Util.kt:
```diff
@@ -0,0 +1,2209 @@
+            val bais =
+                ByteArrayInputStream(resultingCertBytes)
+            cf.generateCertificate(bais) as X509Certificate
+        } catch (e: IOException) {
+            throw IllegalStateException("Error signing key with private key", e)
+        } catch (e: InvalidKeyException) {
+            throw IllegalStateException("Error signing key with private key", e)
+        } catch (e: KeyStoreException) {
+            throw IllegalStateException("Error signing key with private key", e)
+        } catch (e: NoSuchAlgorithmException) {
+            throw IllegalStateException("Error signing key with private key", e)
+        } catch (e: SignatureException) {
+            throw IllegalStateException("Error signing key with private key", e)
+        } catch (e: UnrecoverableEntryException) {
+            throw IllegalStateException("Error signing key with private key", e)
+        } catch (e: CertificateException) {
+            throw IllegalStateException("Error signing key with private key", e)
+        }
```
join catches

davidz25 commented:
Took all the good suggestions up-until `Util.kt` and force-pushed the branch with those.

davidz25 commented:
Also, some unit tests didn't compile and when they did, they failed (there was a one-off in one of the kotlinized low-level utils, fixed that). Added a commit with fixes for this and also manually tested all apps and samples. Works fine.
